ddtrace/sampler.py

```diff
@@ -44,59 +46,67 @@ def set_sample_rate(self, sample_rate):
+        self._by_service_samplers[self._default_key] = RateSampler(sample_rate)
 
-    def __init__(self, tps):
-        self.buffer_limit = tps * self.BUFFER_DURATION
+    def _key(self, service="", env=""):
+        service = service or ""
+        env = env or ""
+        return "service:" + service + ",env:" + env
 
-        # Circular buffer counting sampled traces over the last `BUFFER_DURATION`
-        self.counter = 0
-        self.counter_buffer = array.array('L', [0] * self.BUFFER_SIZE)
-        self._buffer_lock = threading.Lock()
-        # Last time we sampled a trace, multiplied by `BUCKETS_PER_S`
-        self.last_track_time = 0
+    def _set_sample_rate_by_key(self, sample_rate, key):
+        with self._lock:
+            if key in self._by_service_samplers:
+                self._by_service_samplers[key].set_sample_rate(sample_rate)
+            else:
+                self._by_service_samplers[key] = RateSampler(sample_rate)
 
-        log.info("initialized ThroughputSampler, sample up to %s traces/s", tps)
+    def set_sample_rate(self, sample_rate, service="", env=""):
+        self._set_sample_rate_by_key(sample_rate, self._key(service, env))
 
     def sample(self, span):
-        now = int(span.start * self.BUCKETS_PER_S)
-
-        with self._buffer_lock:
-            last_track_time = self.last_track_time
-            if now > last_track_time:
-                self.last_track_time = now
-                self.expire_buckets(last_track_time, now)
-
-            span.sampled = self.counter < self.buffer_limit
-
-            if span.sampled:
-                self.counter += 1
-                self.counter_buffer[self.key_from_time(now)] += 1
-
-        return span
-
-    def key_from_time(self, t):
-        return t % self.BUFFER_SIZE
-
-    def expire_buckets(self, start, end):
-        period = min(self.BUFFER_SIZE, (end - start))
-        for i in range(period):
-            key = self.key_from_time(start + i + 1)
-            self.counter -= self.counter_buffer[key]
-            self.counter_buffer[key] = 0
+        tags = span.tracer().tags
+        env = tags['env'] if 'env' in tags else None
+        key = self._key(span.service, env)
+        with self._lock:
+            if key in self._by_service_samplers:
+                return self._by_service_samplers[key].sample(span)
+            return self._by_service_samplers[self._default_key].sample(span)
+
+    def set_sample_rates_from_json(self, body):
+        log.debug("setting sample rates from JSON '%s'" % repr(body))
+        try:
+            if not isinstance(body, str):
+                body = body.decode('utf-8')
+            if body.startswith('OK'):
+                # This typically happens when using a priority-sampling enabled
+                # library with an outdated agent. It still works, but priority sampling
+                # will probably send too many traces, so the next step is to upgrade agent.
```

Comment:
There @palazzem when we're here we could downgrade and disable priority sampling and switch on protocol v3. OTOH it introduces coupling between components and some complexity I think.